### PR TITLE
Update to support 31402 to 31800 conditional serialization.

### PR DESCRIPTION
**Pull request by pmienk**

### CODE REVIEW

File: src/chain/header.cpp [hunk 1]
```diff
@@ -224,7 +230,22 @@ uint64_t header::serialized_size(bool with_transaction_count) const
...
+    mutex_.lock_upgrade();
+
+    if (hash_ == nullptr)
+    {
+        mutex_.unlock_upgrade_and_lock();
+
+        if (hash_ == nullptr)
+            hash_.reset(new hash_digest(bitcoin_hash(to_data(false))));
+
+        mutex_.unlock();
```
Comment (evoskuil) [inline src/chain/header.cpp]: This needs to be `mutex_.unlock_and_lock_upgrade();`

File: src/chain/transaction.cpp [hunk 2]
```diff
@@ -263,7 +269,22 @@ std::string transaction::to_string(uint32_t flags) const
...
+    mutex_.lock_upgrade();
+
+    if (hash_ == nullptr)
+    {
+        mutex_.unlock_upgrade_and_lock();
+
+        if (hash_ == nullptr)
+            hash_.reset(new hash_digest(bitcoin_hash(to_data())));
+
+        mutex_.unlock();
```
Comment (evoskuil) [inline src/chain/transaction.cpp]: Same here.

**Comment by pmienk** [comment]
Thanks for the catch, corrected.

**Comment by evoskuil** [comment]
I've rebased this on top of my last merged PR and added a patch, see:  #431

**Pull request closed without merging**